egs/wsj/s5/utils/lang/make_lexicon_fst.py

```diff
@@ -0,0 +1,407 @@
+    for (word, pronprob, pron) in lexicon:
+        cost = -math.log(pronprob)
+        cur_state = loop_state
+        for i in range(len(pron) - 1):
+            print("{src}\t{dest}\t{phone}\t{word}\t{cost}".format(
+                src=cur_state,
+                dest=new_state,
+                phone=pron[i],
+                word=(word if i == 0 else '<eps>'),
+                cost=(cost if i == 0 else 0.0)))
+            cur_state = new_state
+            new_state += 1
+
+        i = len(pron) - 1  # note: i == -1 if pron is empty.
+        print("{src}\t{dest}\t{phone}\t{word}\t{cost}".format(
+            src=cur_state,
+            dest=loop_state,
+            phone=(pron[i] if i >= 0 else '<eps>'),
+            word=(word if i <= 0 else '<eps>'),
+            cost=(cost if i <= 0 else 0.0)))
+
+    if nonterminals is not None:
+        new_state = write_nonterminal_arcs(
+            start_state, loop_state, new_state,
+            nonterminals, left_context_phones)
+
+    print("{state}\t{final_cost}".format(
+        state=loop_state,
+        final_cost=0.0))
+
+
+def write_fst_with_silence(lexicon, sil_prob, sil_phone, sil_disambig,
+                           nonterminals=None, left_context_phones=None):
+    """Writes the text format of L.fst to the standard output.  This version is for
+       when --sil-prob != 0.0, meaning there is optional silence
+     'lexicon' is a list of 3-tuples (word, pron-prob, prons)
+         as returned by read_lexiconp().
+     'sil_prob', which is expected to be strictly between 0.. and 1.0, is the
+         probability of silence
+     'sil_phone' is the silence phone, e.g. "SIL".
+     'sil_disambig' is either None, or the silence disambiguation symbol, e.g. "#5".
+     'nonterminals', which relates to grammar decoding (see kaldi-asr.org/doc/grammar.html),
+        is either None, or the user-defined nonterminal symbols as a list of
+        strings, e.g. ['#nonterm:contact_list', ... ].
+     'left_context_phones', which also relates to grammar decoding, and must be
+        supplied if 'nonterminals' is supplied is either None or a list of
+        phones that may appear as left-context, e.g. ['a', 'ah', ... '#nonterm_bos'].
+    """
+
+    assert sil_prob > 0.0 and sil_prob < 1.0
+    sil_cost = -math.log(sil_prob)
+    no_sil_cost = -math.log(1.0 - sil_prob);
+
+    start_state = 0
+    loop_state = 1  # words enter and leave from here
+    sil_state = 2   # words terminate here when followed by silence; this state
+                    # has a silence transition to loop_state.
+    new_state = 3  # the next un-allocated state, will be incremented as we go.
+
+
+    print('{src}\t{dest}\t{phone}\t{word}\t{cost}'.format(
+        src=start_state, dest=loop_state,
+        phone='<eps>', word='<eps>', cost=no_sil_cost))
+    print('{src}\t{dest}\t{phone}\t{word}\t{cost}'.format(
+        src=start_state, dest=sil_state,
+        phone='<eps>', word='<eps>', cost=sil_cost))
```

Comment:
I guess the cost from start_state to sil_state should be 0.0 here or it would add sil_cost twice if the sentence starts with SIL.